commit e3bf6799983eea213f33266d60a44428d3b19da5
Author: Norman Maurer <norman_maurer@apple.com>
Date:   Thu Mar 10 14:58:46 2016 +0100

    Ensure connect promise is notifed before fireChannelActive() is called. Related to [#4927]
    
    Motivation:
    
    We should always first notify the promise before trigger an event through the pipeline to be consistent.
    
    Modifications:
    
    Ensure we notify the promise before fire event.
    
    Result:
    
    Consistent behavior

diff --git a/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDatagramChannel.java b/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDatagramChannel.java
index e540328..5c4b045 100644
--- a/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDatagramChannel.java
+++ b/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDatagramChannel.java
@@ -500,6 +500,9 @@ public final class EpollDatagramChannel extends AbstractEpollChannel implements
                     EpollDatagramChannel.this.local = fd().localAddress();
                     success = true;
 
+                    // First notify the promise before notifying the handler.
+                    channelPromise.trySuccess();
+
                     // Regardless if the connection attempt was cancelled, channelActive() event should be triggered,
                     // because what happened is what happened.
                     if (!wasActive && isActive()) {
@@ -509,12 +512,11 @@ public final class EpollDatagramChannel extends AbstractEpollChannel implements
                     if (!success) {
                         doClose();
                     } else {
-                        channelPromise.setSuccess();
                         connected = true;
                     }
                 }
             } catch (Throwable cause) {
-                channelPromise.setFailure(cause);
+                channelPromise.tryFailure(cause);
             }
         }
 
